### PR TITLE
Add S3 support

### PR DESCRIPTION
**Pull request by lolejar**
- add S3 support for restoring from S3 repositories
- add S3 configuration as a code

### CODE REVIEW

**Comment by samrocketman** [comment]
Thanks for contributing I’ll take a look and test as soon as I can.

**Review by samrocketman — APPROVED** [review]
Code looks good!  I’m not able to test this right now but I will be merging this after I do.

File: docs/repositories.md [hunk 1]
```diff
@@ -25,14 +25,33 @@ format validation.
...
         "file": [
             "default"
+        ],
+        "s3": [
+                {
+                    "name": "s3-test-blob",
+                    "config": {
+                        //mandatory fields
+                        "bucket" : "bucket-name",
+                        "accessKeyId": "your access key id",
```
Comment (dawidmalina) [inline docs/repositories.md:43]: It could be also optional fields as you can use iam roles and then you don't need to specify access and secret key.

Comment (zxkane) [reply docs/repositories.md:43]: The comment in line 41 already points out the only mandatory field `bucket`.

Any plan to merge it or still require follow up changes?

Comment (zxkane) [reply docs/repositories.md:43]: sorry for noise. Agree with @dawidmalina comments

File: functions/repositoryConfiguration.groovy [hunk 2]
```diff
@@ -209,14 +209,14 @@ void validateConfiguration(def json) {
...
+        if((!(json['blobstores']?.get('file') in List) || false in json['blobstores']?.get('file').collect { it in String }) && (!(json['blobstores']?.get('s3') in List) || false in json['blobstores']?.get('s3').collect { it.name in String && it.config.bucket in String && it.config.accessKeyId in String && it.config.secretAccessKey in String && it.config.expiration in String})) {
+            println "blobstore file type must contain a list of Strings."
         }
         checkForEmptyValidation('repository providers', ((json['repositories']?.keySet() as List) - supported_repository_providers))
         checkForEmptyValidation('repository types', (json['repositories'].collect { k, v -> v.keySet() as List }.flatten().sort().unique() - supported_repository_types))
         checkForEmptyValidation('blobstores defined in repositories.  The following must be listed in the blobstores',
-                (getKnownDesiredBlobStores(json) - json['blobstores']['file']))
-        checkRepositorFormat(json)
+                (getKnownDesiredBlobStores(json) - json['blobstores']['file'] - json['blobstores']['s3'].name))
```
Comment (zxkane) [inline functions/repositoryConfiguration.groovy:218]: `json['blobstores']['s3'].name` will cause NPE if no `s3` defined in configuration file.

**Pull request merged**